### PR TITLE
Excerpt strip fields.

### LINKED ISSUE #7758: [RFC] Exerpt runtime strip keys
Hello,

`Bolt\Helpers\Excerpt` is great but it could be improved by adding `$stripKeys` as a fourth argument. This way, one could easily remove unwanted fields from a search result.

At the moment, the `$stripKeys` array is hardcoded [here](https://github.com/bolt/bolt/blob/3.6/src/Helpers/Excerpt.php#L58), a simple `array_merge` would do the job. An optional fourth argument wouldn't break any BC.

Keen on PR if this is approved 🙂

### PR DESCRIPTION
**Pull request by artggd**
Fixes #7758

Allow to strip unwanted fields out of the excerpt by adding a fourth argument to Excerpt and related runtime methods.

Doc update on its way.

### CODE REVIEW

**Comment by JarJak** [comment]
Could you add a test for it?

**Comment by artggd** [comment]
Sure thing, fixing checks first :)

By the way, the CI fails because of a bad call of `content|excerpt` in `bolt/_base/_listing.twig` where `content.excerpt` should've been called.

`content.excerpt` and `content|excerpt` don't take the same arguments (`.excerpt` has `$includeTitle` as second argument). This didn't break in the past because the last argument was simply ignored. Now that I added another one, type hinting breaks.

Although this PR doesn't break BC, I'm afraid it could be dangerous in the end...

**Review by bobdenotter — DISMISSED** [review]
> content.excerpt and content|excerpt don't take the same arguments (.excerpt has $includeTitle as second argument). 

This is because the `.excerpt` works on a `content` object only, but `|excerpt` is a generic filter, that works on strings, arrays or content-like objects. 

> Although this PR doesn't break BC, I'm afraid it could be dangerous in the end...

It seems to work fine, and I don't spot immediate issues.. 👍 

Are you worried about merging it in?

**Comment by artggd** [comment]
Well, I understand the difference between `.excerpt` and `|excerpt` but it's easily confusable. Taking for proof there was a miscall in Bolt's core.

I reckon there might be quite an amount of projects miscalling `|excerpt` and passing the same args as `.excerpt`. Those will break on their next `composer up`.

Though I really need this feature, I'd say the merge could wait for the next major release. Unless you have any way of measuring the risk and/or are willing to accept it.

Your call ;)

**Comment by artggd** [comment]
Or, maybe we could remove the type hint of `$stripFields` so that if a fourth arg is wrongly passed to `.excerpt` (miscall of `$focus` (string)) it won't break. But I'm not fond of this option, hacky...

**Comment by JarJak** [comment]
Actually $stripFields param could be added only to Content method, because it is valid only with Content context.
I would leave excerpt filter as-is, then nothing will break for anyone :)

Docs needs an update, can you provide separate PR for them @artggd ?

**Comment by artggd** [comment]
The filter is very versatile and is valid for `string`, `Content` or `array`. It still makes sense to have `$stripFields` in both filter and Content method IMO. 
The issue would remain anyway, it breaks when Content method is called with the filter params, not the other way round. Which is probably the most common case.

I'm thinking removing type hinting could actually be an ok solution with proper type control in the filter. It would also add some flexibility, allowing for usage of the filter like so 
```TWIG
{{ stuff|excerpt(stripFields="strip out those fields") }}
```
 Just need to explode `$stripFields` when it's a string.

If you guys are happy with this option, I'll make the update and add a couple of tests.

@JarJak I did provide a PR for doc update 😉👉 https://github.com/bolt/docs/pull/953

**Comment by JarJak** [comment]
>The filter is very versatile and is valid for string, Content or array. 

Yes, you can use filter on content, but this is more like extra feature than normal usage. And since $stripFields makes sense only in Content context, I would leave this parameter only in Content method.

With that being said, IMO $stripKeys should also be moved into Content method, but that would introduce BC break.

However I won't block this PR because of my opinion :)
@bobdenotter what do you think?

**Review by JarJak — DISMISSED** [review]

**Comment by JarJak** [comment]
Oh, and I'm not happy with
```
{{ stuff|excerpt(stripFields="strip out those fields") }}
```
As an option. It doesn't make sense to me to allow such way of using this feature.

**Comment by artggd** [comment]
Agreed, if we want to restrict $stripFields on Content only, then let's move out all the related logic to Content and leave the filter to only work with strings.  It'd be way more consistent.

What about merging this as a transitional step and keep strict typing for next major?

**Comment by artggd** [comment]
> Oh, and I'm not happy with
> 
> ```
> {{ stuff|excerpt(stripFields="strip out those fields") }}
> ```
> As an option. It doesn't make sense to me to allow such way of using this feature.

Then $stripFields could simply be ignored if string and raise a warning (rather than a 500)

**Comment by JarJak** [comment]
@artggd in next major (Bolt 4) we made a slightly different approach. Fields that are allowed in excerpt must implement Excerptable interface. Look for usages of https://github.com/bolt/four/blob/master/src/Entity/Field/Excerptable.php 
And as always, PRs are more than welcome!

**Comment by artggd** [comment]
@JarJak This sounds very nice 🙂
Although I can see field values are stored in a json column, I don't know how you're going to achieve search without pulling out heavy indexing!
Digressing, this is not the place to discuss this but I'd be more than happy to help. I've had similar issue in the past and solved it with a different approach but I guess Bolt 4's architecture is already thought and written down.

**Comment by JarJak** [comment]
@artggd we are open for all ideas and we can change everything before beta :)

**Comment by artggd** [comment]
@JarJak I'll catch-up with you on https://github.com/bolt/four/issues/306 when I get some time :)

Short recap of this issue, we've got 4 options left:
1. Merging as is.
1. Removing `$stripFields` type hint and raise a warning if not an array.
1. Allowing `$stripFields` on Content method only and move the Content related logic out of the filter (inducing a BC break).
1. Allowing `$stripFields` on Content method only and keep Content related logic in the filter.

My personal fav being number 2

**Comment by JarJak** [comment]
I'm ok with either 2 or 4. @bobdenotter ?

**Comment by bobdenotter** [comment]
Nr. 2 has my preference! 👍

**Comment by artggd** [comment]
Sweet then, working on it 👍

**Comment by artggd** [comment]
Hey so, I'm still working but I've ran into a small coding issue.

I decided to log a message rather than '@trigger_error' because trigger_error doesn't show up in the debug toolbar unless it's a deprecated call. Also, it's a pain to unit test.

Problem is, `Excerpt` is not a service so I have to duplicate the same bit of code in the callers (which are services and can access the logger). The said bit of code tests the type of $stripFields and logs an error.

I'm really not happy with that.

Any recommendation? How'd you achieve that in a cleaner way?

**Comment by JarJak** [comment]
Normally I would just throw an Exception there.

**Comment by artggd** [comment]
I would too, but that was the point Nr. 2. Not throwing an exception so it doesn't break in case of a miscall 🙂

**Comment by JarJak** [comment]
Then I think we can deprecate it.

Decorating every excerpt call with logger is not an option because there can always be n+1 call in extension or user's code.

File: src/Helpers/Excerpt.php
Comment (JarJak) [inline src/Helpers/Excerpt.php]: default value triggers error?

**Review by JarJak — CHANGES_REQUESTED** [review]

File: src/Helpers/Excerpt.php
Comment (artggd) [reply src/Helpers/Excerpt.php]: I realize that this morning, update coming up now

**Comment by artggd** [comment]
It should be ready to merge, I also added 3 tests on ContentTests.

File: src/Helpers/Excerpt.php
Comment (JarJak) [inline src/Helpers/Excerpt.php]: just set default value of $sripFields to [] here

**Review by JarJak — CHANGES_REQUESTED** [review]

File: src/Helpers/Excerpt.php
Comment (JarJak) [inline src/Helpers/Excerpt.php]: remove null

**Review by JarJak — CHANGES_REQUESTED** [review]

**Comment by artggd** [comment]
Not sure what's wrong with the CI, the tests run smooth on my machine and it seems to fail on nothing related to my last dev.

**Review by JarJak — APPROVED** [review]

**Comment by bobdenotter** [comment]
Tests are failing, a lot of: 

```
Error: Call to undefined method Twig\TemplateWrapper::getTemplateName()
```

Looks like this is introduced in this update: https://symfony.com/blog/twig-sandbox-information-disclosure

**Comment by bobdenotter** [comment]
The issue with Twig is fixed. See: https://github.com/twigphp/Twig/issues/2892

**Comment by artggd** [comment]
Great! Thanks @bobdenotter

**Comment by bobdenotter** [comment]
@artggd Thank _you_ for the PR! 👍

**Pull request merged**
